dev_tools/nyaml2nxdl/nyaml2nxdl.py

```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env python3
+"""Main file of nyaml2nxdl tool.
+Users create NeXus instances by writing a YAML file
+which details a hierarchy of data/metadata elements
+
+"""
+# -*- coding: utf-8 -*-
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import os
+import xml.etree.ElementTree as ET
+
+import click
+
+from .nyaml2nxdl_backward_tools import Nxdl2yaml
+from .nyaml2nxdl_backward_tools import compare_niac_and_my
+from .nyaml2nxdl_forward_tools import nyaml2nxdl
+from .nyaml2nxdl_forward_tools import pretty_print_xml
+from .nyaml2nxdl_helper import extend_yamlfile_with_comment
+from .nyaml2nxdl_helper import get_sha256_hash
+from .nyaml2nxdl_helper import separate_hash_yaml_and_nxdl
+
+DEPTH_SIZE = 4 * " "
+
+# NOTE: Some handful links for nyaml2nxdl converter:
+# https://manual.nexusformat.org/nxdl_desc.html?highlight=optional
+
+
+def generate_nxdl_or_retrieve_nxdl(yaml_file, out_xml_file, verbose):
+    """
+    Generate yaml, nxdl and hash.
+    if the extracted hash is exactly the same as producd from generated yaml then
```

Comment:
Insert one blank line before this one.  This is standard practice.  First line is one sentence, all on one line.  If more documentation is provided, then a blank line and the additional documentation follows.

Reply:
:+1: